internal/pkg/vm/internal/vmcontext/testing.go

```diff
@@ -0,0 +1,395 @@
+		return err
+	}
+	return nil
+}
+
+//
+// Applier
+//
+
+type ValidationApplier struct{}
+
+func (a *ValidationApplier) ApplyMessage(context *vtypes.ExecutionContext, state vstate.VMWrapper, msg *vtypes.Message) (vtypes.MessageReceipt, error) {
+	st := state.(*ValidationVMWrapper)
+
+	// set epoch
+	// Note: this would have normally happened during `ApplyTipset()`
+	st.vm.currentEpoch = context.Epoch
+
+	// map message
+	// Dragons: fix after cleaning up our msg
+	ourmsg := &types.UnsignedMessage{
+		To:         msg.To,
+		From:       msg.From,
+		CallSeqNum: uint64(msg.CallSeqNum),
+		Value:      msg.Value,
+		Method:     msg.Method,
+		Params:     msg.Params,
+		GasPrice:   msg.GasPrice,
+		GasLimit:   types.GasUnits(msg.GasLimit.Uint64()),
+	}
+
+	// invoke vm
+	ourreceipt, _, _ := st.vm.applyMessage(ourmsg, ourmsg.OnChainLen(), context.MinerOwner)
+
+	// commit and persist changes
+	// Note: this is not done on production for each msg
+	if err := st.PersistChanges(); err != nil {
+		return vtypes.MessageReceipt{}, err
+	}
+
+	// map receipt
+	receipt := vtypes.MessageReceipt{
+		ExitCode:    ourreceipt.ExitCode,
+		ReturnValue: ourreceipt.ReturnValue,
+		GasUsed:     big.Int(ourreceipt.GasUsed),
+	}
+
+	return receipt, nil
+}
+
+//
+// KeyManager
+//
+
+type KeyManager struct {
+	// Private keys by address
+	keys map[address.Address]*gfcrypto.KeyInfo
+
+	// Seed for deterministic secp key generation.
+	secpSeed int64
+	// Seed for deterministic bls key generation.
+	blsSeed int64 // nolint: structcheck
+}
+
+func newKeyManager() *KeyManager {
+	return &KeyManager{
+		keys:     make(map[address.Address]*gfcrypto.KeyInfo),
+		secpSeed: 0,
+	}
+}
+
+func (k *KeyManager) NewSECP256k1AccountAddress() address.Address {
+	secpKey := k.newSecp256k1Key()
+	addr, err := secpKey.Address()
+	if err != nil {
+		panic(err)
+	}
+	k.keys[addr] = secpKey
+	return addr
+}
+
+func (k *KeyManager) NewBLSAccountAddress() address.Address {
+	blsKey := k.newBLSKey()
+	addr, err := blsKey.Address()
+	if err != nil {
+		panic(err)
+	}
+	k.keys[addr] = blsKey
+	return addr
+}
+
+func (k *KeyManager) Sign(addr address.Address, data []byte) (crypto_spec.Signature, error) {
+	ki, ok := k.keys[addr]
+	if !ok {
+		return crypto_spec.Signature{}, fmt.Errorf("unknown address %v", addr)
+	}
+	switch ki.Type() {
+	case crypto_spec.SigTypeBLS:
+		var pk ffi.PrivateKey
+		copy(pk[:], ki.PrivateKey)
+		digest := ffi.PrivateKeySign(pk, data)
+		return crypto_spec.Signature{
+			Type: crypto_spec.SigTypeBLS,
+			Data: digest[:],
+		}, nil
+	case crypto_spec.SigTypeSecp256k1:
+		b2sum := blake2b.Sum256(data)
+		digest, err := crypto.Sign(ki.PrivateKey, b2sum[:])
+		if err != nil {
+			return crypto_spec.Signature{}, err
+		}
+		return crypto_spec.Signature{
+			Type: crypto_spec.SigTypeSecp256k1,
+			Data: digest,
+		}, nil
+	default:
+		return crypto_spec.Signature{}, fmt.Errorf("unknown key type: %v for address: %s", ki.Type(), addr)
+	}
```

Comment:
Much of this code is probably duplicating stuff in our crypto package (which improved recently). Use crypto/signatures.go if possible.